Usable server and client

dgalaktionov opened the pull request:
Hello again! I've added some listeners for `Server`, created a convenient `Client` class and, more importantly, made `Connection` handle all the negotiation with ideas from [gafferongames](https://web.archive.org/web/20181107181442/https://gafferongames.com/post/client_server_connection/):
1. Client sends a negotiation packet to server, with its public key and a randomly generated challenge code.
2. Server replies with its negotiation code, public key and another random challenge code.
3. Since the IV are now initialized, Client sends a confirmation packet with encrypted `client_challenge^server_challenge`. From this point, the client considers itself connected and stops sending negotiation packets, although it keeps replying with a confirmation for every time it gets a negotiation packet from the server, as the older confirmation may have been lost.
4. The server decrypts and validates the challenge code and also moves to connected state, so the negotiation is over for both.

Apart from the tests, you can check how the new model is supposed to work with [my example client and server programs](https://gist.github.com/dgalaktionov/4f5602082d9f9168fae2d46b18943919). I've also used [clumsy](https://jagt.github.io/clumsy/) (what a wonderful tool) to ensure that the connection is robust and eventually establishes under unstable networks with lost packets, unless the situation is very bad and the connection times out.

I understand this is a rather large PR and may take some time to review, but it would feel incomplete if I made it any smaller. Nevertheless, I'd like to have you input on some of my concerns when you have the time:
- Do you see any problems with the negotiation method? Am I correct in my understanding that ChaCha20 is considered secure against known plaintext attacks?
- A lot of the code looks the same in `Client` and `Server`. I could include the listeners in ICommunication and let `Connection` grow a bit larger by handing most of the logic from `Client::ProcessPacket` and `Server::ProcessPacket`. What do you think?
- As it is now, memory is allocated before every packet sent, and freed after. I'm not sure how much performance overhead that implies. I guess someone is working on #1 so we could find out soon...

dgalaktionov committed:
IPv4 endpoints support for Server::Send

dgalaktionov committed:
Make server aware of connection negotiation

Things left to do:
- Make a Client that also has connection negotiation
- Extendable listeners
- Handle disconnecting clients
- Tests!

dgalaktionov committed:
Negotiation between client and server

Introduced challenge codes. It's unsecure and only useful for small tests.

Also introduced kPayload header type to differentiate it from kConnection, which is now used for client challenge response.

dgalaktionov committed:
Error codes

Instead of cryptic bools

dgalaktionov committed:
Secure challenge

Use a two way code exchange, xor and the crypto filter.

TIL how ChaCha20 works.

dgalaktionov committed:
Server & client listeners

TODO: handle disconnects

dgalaktionov committed:
Robust handling of disconnections

dgalaktionov committed:
Move Client code to a source file

dgalaktionov committed:
Tests for new client and server

dgalaktionov committed:
Fix StackAllocator declared too small for Buffer

For some reason, only segfaults on linux

dgalaktionov committed:
Fix formatting

maximegmd commented on Code/network/include/Connection.h:
```diff
@@ -38,15 +40,19 @@ class Connection
     };
 
-    Connection(ICommunication& aCommunicationInterface, const Endpoint& acRemoteEndpoint);
+    Connection(ICommunication& aCommunicationInterface, const Endpoint& acRemoteEndpoint, const bool acIsServer=false);
```
We don't pass const by value for primitive types

maximegmd commented on Code/network/include/Connection.h:
```diff
@@ -55,22 +61,30 @@ class Connection
-    bool ProcessPacket(Buffer* apBuffer);
-    bool ProcessNegociation(Buffer* apBuffer);
-
+    Outcome<uint64_t, Connection::HeaderErrors> ProcessPacket(Buffer::Reader & aReader);
```
uint64_t doesn't explain what it contains, you can typedef something with an explicit name and/or comment what it does

maximegmd commented on Code/network/include/Connection.h:
```diff
@@ -55,22 +61,30 @@ class Connection
+    uint64_t Update(uint64_t aElapsedMilliseconds);
+    void Disconnect();
+
+    void WriteHeader(Buffer::Writer& aWriter, const uint64_t acHeaderType);
```
const

maximegmd commented on Code/network/include/Connection.h:
```diff
@@ -55,22 +61,30 @@ class Connection
     const Endpoint& GetRemoteEndpoint() const;
 
-    void Update(uint64_t aElapsedMilliseconds);
+    uint64_t Update(uint64_t aElapsedMilliseconds);
```
uint64_t

maximegmd commented on Code/network/include/Connection.h:
```diff
@@ -79,4 +93,7 @@ class Connection
     uint64_t m_timeSinceLastEvent;
     Endpoint m_remoteEndpoint;
     DHChachaFilter m_filter;
+    bool m_isServer;
```
Kinda sucks to store all this stuff for alive connections, might be best to have two other classes, ClientNegociation and ServerNegociation that would do this and then create a Client

dgalaktionov replied:
Sure, but if we want our disconnection packets authenticated we'd still have to store a code per connection...

maximegmd commented on Code/network/include/Server.h:
```diff
@@ -11,19 +11,23 @@ class Server : public AllocatorCompatible
-    bool Start(uint16_t aPort);
-    uint32_t Update(uint64_t aElapsedMilliSeconds);
-    uint16_t GetPort() const;
+    bool Start(const uint16_t acPort) noexcept;
```
const stuff

maximegmd commented on Code/network/src/Client.cpp:
```diff
@@ -0,0 +1,105 @@
+    Buffer::Writer writer(pBuffer);
+    m_connection.WriteHeader(writer, Connection::Header::kPayload);
+    writer.WriteBytes(apData, aLength);
+    bool sent = Send(m_connection.GetRemoteEndpoint(), *pBuffer);
```
UDP is limited by an MTU, there needs to be a mechanism for splitting and reassembling if we are over the capacity

dgalaktionov replied:
Ofc I'm aware. I was planning on handling fragmentation in another PR. I haven't thought of a solution I'm satisfied with yet, so it may take a couple of days.

maximegmd commented on Code/network/src/Connection.cpp:
```diff
@@ -126,28 +222,77 @@ void Connection::Update(uint64_t aElapsedMilliseconds)
+{
     StackAllocator<1 << 13> allocator;
-    auto* pBuffer = allocator.New<Buffer>(1200);
+    auto* pBuffer = allocator.New<Buffer>(m_isServer ? 200 : 1200);
```
It's allocated on the stack, don't need to optimize the size

dgalaktionov replied:
This is not about optimization, but to prevent the server being useful for reflection in a DDOS attack (see error `kPayloadRequired`). The gafferon article I've linked also speaks about amplification, but if the packets sizes are the same there's no real amplification going on.

dgalaktionov replied:
BTW it's not actually stack allocated, as `Buffer` still uses its default `StandardAllocator` internally for `m_pData`. Want me to fix that? I could give it an optional `apAllocator` parameter in the constructor.

maximegmd commented on Code/network/src/Server.cpp:
```diff
@@ -13,62 +13,132 @@ Server::~Server()
-    m_connectionManager.Update(aElapsedMilliSeconds);
-
-    return Work();
+    uint32_t nPackets = Work();
```
nPackets is not explicit, packetCount would be better

maximegmd commented on Code/network/src/Server.cpp:
```diff
@@ -13,62 +13,132 @@ Server::~Server()
-
-    return Work();
+    uint32_t nPackets = Work();
+    m_connectionManager.Update(acElapsedMilliSeconds, std::bind(&Server::OnClientDisconnected, this, std::placeholders::_1));
```
Use a lambda, bind is very outdated

maximegmd commented on Code/tests/src/network.cpp:
```diff
@@ -326,38 +287,265 @@ TEST_CASE("Connection", "[network.connection]")
+            }
+        }
+
+        size_t NumClients()
```
has to a be a getter and const

maximegmd reviewed (changes requested):
I believe Client is a bit too early right now, we don't have the middle step being packet fragmentation, channel management, reliability and reassembly of fragments

dgalaktionov committed:
Code quality fixes

- Remove const from primitive types
- Use a HeaderType and Connection::State where appropriate instead of uint64_t
- Use lambda instead of std::bind

dgalaktionov committed:
Fix mandatory padding for client

I have overlooked the fact that the receiver's buffer is always of 1200 bytes. In order to force the client to use a padding, I'm making it skip 1000 bytes and write at the end of its negotiation packets instead.

Also replaced these magic numbers by actual constants.

dgalaktionov committed:
Fix unsafe possible fallthrough in switch

dgalaktionov committed:
A very basic Message class

No fragmentation yet

dgalaktionov committed:
A bit more complex Message

Can be internally partitioned in slices. All that is left for message fragmentation & reconstruction is a message merging method.

dgalaktionov committed:
Message fragmentation

Plus one extremely mean test case

dgalaktionov committed:
MessageReceiver

With a buffer to reassemble the last fragmented messages

dgalaktionov committed:
Fragmentation for client and server

TODO: refactor duplicated code

dgalaktionov commented:
To add support for fragmentation, I've added a `Message` class in protocol. After much thought, I've decided to avoid letting `Message` know anything about the network. This implies that it has no way of telling how many fragments it is going to create in advance, as it does not know what is the maximum packet size. Instead, for each fragment I'm writing a message sequence number, the fragment offset byte and the fragment length.

These fragments are sent with no reliability at all, and I really don't consider that reliable channels fall in the scope of this PR. If you think my fragmentation strategy is solid, in the future we could implement it with an additional per-packet sequence and ACK. Then the sender would have to keep track of the reliable parts of the last sent packets and keep resending those parts until the reliable queue or buffer can be cleared.

I also don't quite understand what did you mean by `ClientNegociation` and `ServerNegociation`, so I didn't make any attempt to refactor `Connection`. In fact, I made it grow a bit larger, as we must store a buffer for reassembly one way or another... Although I'm still concerned about the large amount of duplicated code between `Client` and `Server`, it calls for some refactoring.

Take your time to reply, as I don't expect to have much free time until next week...
